### ISSUE TITLE
path option for info

### ISSUE THREAD

**Issue by stuartthebruce** [zevv/duc#254]
It would be helpful for "duc info" to accept an optional path to display statistics starting from that directory.

The closest I can currently find is "duc xml | head", however, that does not give separate numbers for files and directories, e.g.,
```
# duc info -bd ./home2.duc
Date       Time       Files    Dirs    Size Path
2020-07-15 17:00:05 652925601 50060784 312747364261888 /home2

# duc xml -d home2.duc /home2 | head -2
<?xml version="1.0" encoding="UTF-8"?>
<duc root="/home2" size_apparent="379730437625610" size_actual="312747364245504" count="702986384">

# duc xml -d home2.duc /home2/jacob.lange | head -2
<?xml version="1.0" encoding="UTF-8"?>
<duc root="/home2/jacob.lange" size_apparent="24796967303659" size_actual="28332256509952" count="126440122">
```

and what I would like is,
```
# duc info -bd ./home2.duc /home2/jacob.lange
Date       Time       Files    Dirs    Size Path
2020-07-15 17:00:05 126321718 285303 28332256509952 /home2/jacob.lange
```

**Comment by l8gravely** [comment]
stuartthebruce> It would be helpful for "duc info" to accept an
stuartthebruce> optional path to display statistics starting from that
stuartthebruce> directory.

Don't you want 'du ls -d home2.duc /home/2/jacob.lange' instead?

stuartthebruce> The closest I can currently find is "duc xml | head", however, that does not give separate numbers
stuartthebruce> for files and directories, e.g.,

stuartthebruce> # duc info -bd ./home2.duc
stuartthebruce> Date       Time       Files    Dirs    Size Path
stuartthebruce> 2020-07-15 17:00:05 652925601 50060784 312747364261888 /home2

stuartthebruce> # duc xml -d home2.duc /home2 | head -2
stuartthebruce> <?xml version="1.0" encoding="UTF-8"?>
stuartthebruce> <duc root="/home2" size_apparent="379730437625610" size_actual="312747364245504" count="702986384">

stuartthebruce> # duc xml -d home2.duc /home2/jacob.lange | head -2
stuartthebruce> <?xml version="1.0" encoding="UTF-8"?>
stuartthebruce> <duc root="/home2/jacob.lange" size_apparent="24796967303659" size_actual="28332256509952" count="126440122">

stuartthebruce> and what I would like is,

stuartthebruce> # duc info -bd ./home2.duc /home2/jacob.lange
stuartthebruce> Date       Time       Files    Dirs    Size Path
stuartthebruce> 2020-07-15 17:00:05 126321718 285303 28332256509952 /home2/jacob.lange

stuartthebruce> —
stuartthebruce> You are receiving this because you are subscribed to this thread.
stuartthebruce> Reply to this email directly, view it on GitHub, or unsubscribe.*

**Comment by stuartthebruce** [comment]
"duc ls" is indeed very close to what I want. In particular, I can run that 3 times to get a summary of each top-level directory with respect to: size, apparent size, and count. So if you don't like the idea of "duc info path" then as an alternative enhancement are you willing to consider adding a "duc ls" option that distinguishes count between "Files" and "Dirs" the way "duc info" does?

**Comment by l8gravely** [comment]
>>>>> "stuartthebruce" == stuartthebruce  <notifications@github.com> writes:

stuartthebruce> "duc ls" is indeed very close to what I want. In
stuartthebruce> particular, I can run that 3 times to get a summary of
stuartthebruce> each top-level directory with respect to: size,
stuartthebruce> apparent size, and count. So if you don't like the
stuartthebruce> idea of "duc info path" then as an alternative
stuartthebruce> enhancement are you willing to consider adding a "duc
stuartthebruce> ls" option that distinguishes count between "Files"
stuartthebruce> and "Dirs" the way "duc info" does?


duc ls --count /some/path

will show you the file counts.  It's probably not hard to provide a
combined view at some point.

John

**Comment by stuartthebruce** [comment]
Are there any other commands besides "duc info" that distinguish between "Files" and "Dirs"?

**Comment by l8gravely** [comment]
stuartthebruce> Are there any other commands besides "duc info" that
stuartthebruce> distinguish between "Files" and "Dirs"?

Off hand... I don't remember.  duc -h, du <cmd> -h will show you more
details.  And there's always the source. :-)

The man page should also give you details, and if it's not upto date
let me know and I'll push a patch in the next release.

John

**Comment by stuartthebruce** [comment]
I might have missed it, but I don't see a method to report the size, apparent size, number of regular files, and number of directories underneath a specific directory in an index, though "duc info" does report this for the entire index.

**Comment by l8gravely** [comment]
>>>>> "stuartthebruce" == stuartthebruce  <notifications@github.com> writes:

stuartthebruce> I might have missed it, but I don't see a method to
stuartthebruce> report the size, apparent size, number of regular
stuartthebruce> files, and number of directories underneath a specific
stuartthebruce> directory in an index, though "duc info" does report
stuartthebruce> this for the entire index.

Did you look at 'duc ui' and then toggle the size/files info using the
command keys?

John

**Comment by stuartthebruce** [comment]
Yes, I am aware of 'duc ui'. I should have been clearer, but the context of this request is an easily scriptable way to call duc to get this information for an enhanced cgi script. Thanks.

**Comment by l8gravely** [comment]
stuartthebruce> Yes, I am aware of 'duc ui'. I should have been
stuartthebruce> clearer, but the context of this request is an easily
stuartthebruce> scriptable way to call duc to get this information for
stuartthebruce> an enhanced cgi script. Thanks.

Ok, that makes more sense.  If you want to try hacking the CGI
code... that would we awesome.  :-)

**Comment by stuartthebruce** [comment]
For now I am using a wrapper CGI script that filters the output of "duc cgi" (e.g., buttons to switch between usage, apparent usage and file count) as well as making a few other "duc ls" and "duc info" calls to create a click-to-sort table of top-level directories from a set of large index files with each row linking to the corresponding "duc cgi" output.

What do you think about the original idea in this issue, i.e., "duc index path"?

Or adding another option to "duc ls" that gives all the same information (size, apparent size, file count and directory count) in a single pass?

**Comment by l8gravely** [comment]
>>>>> "stuartthebruce" == stuartthebruce  <notifications@github.com> writes:

stuartthebruce> For now I am using a wrapper CGI script that filters
stuartthebruce> the output of "duc cgi" (e.g., buttons to switch
stuartthebruce> between usage, apparent usage and file count) as well
stuartthebruce> as making a few other "duc ls" and "duc info" calls to
stuartthebruce> create a click-to-sort table of top-level directories
stuartthebruce> from a set of large index files with each row linking
stuartthebruce> to the corresponding "duc cgi" output.

stuartthebruce> What do you think about the original idea in this
stuartthebruce> issue, i.e., "duc index path"?

stuartthebruce> Or adding another option to "duc ls" that gives all
stuartthebruce> the same information (size, apparent size, file count
stuartthebruce> and directory count) in a single pass?

I'm going to have to play with the code (when I get time) to see what
can be done here.  But looking at the CGI output, it seems obvious to
put a link into the header block for Files and Size columns to toggle
between views.  The default would be another link to click on to pass
the 'filecnt=1' flag into the CGI script.  

 db.cgi?cmd=index&path=/foo   123gb  db.cgi?cmd=index&filecnt=1&path=/foo

which would then flip so that the 123gb now has a link without the
filecnt part, so it would reset to disk space view.  So that's one
idea at least...

The next step is the showing the data both ways easily.  Again,
haven't looked at the code yet in detail.

John

**Comment by stuartthebruce** [comment]
Thanks, John. Before considering CGI enhancements what do you think about adding the simpler "duc index path" option?

**Comment by l8gravely** [comment]
>>>>> "stuartthebruce" == stuartthebruce  <notifications@github.com> writes:

stuartthebruce> Thanks, John. Before considering CGI enhancements what
stuartthebruce> do you think about adding the simpler "duc index path"
stuartthebruce> option?

Haven't looked at the code at all, so I have no opinion yet.  Been
busy with kid's birthdays and scouting stuff.  :-)

**Comment by l8gravely** [comment]
Hmm... here's a silly suggestion, which is if you *don't* have a ton of user directories below /home2/, you might try:

duc index -d /tmp/home2.db /home2
duc index -d /tmp/home2.db /home2/* 

it will whine a ton about any non-directory information returned.  In my simple quick and dirty tests, you can then do:

duc info -d /tmp/home2.db 

and it will print the info you want down the next level.  But while testing this, and trying to re-index the same base directory into a currently used DB, I get a segfault.  Which now I need to try and debug and figure out what is going wrong here.  I haven't run into this myself since when I index large multiterabyte filesystems, I split them up into seperate NFS mount points that I undex, so I have one DB-per-mount-point (all automounted as needed for clients) and when I re-index, I do it to a new DB and then move the new DB into place.  I never re-index into the same DB.  

But after looking at the code, I think I can do what you want, but it won't be quick unfortunately, since I need to spend some time figuring out the right way to do this, and in a backwardsly compatible way.

Also, which backend DB are you using?  Just give me the output of 'duc --version' please.

John

**Comment by stuartthebruce** [comment]
I had previously been running separate top-level and separate directory scans, but keeping them in different database files. However, I have recently stopped the redundant directory scans to cut back on filesystem metadata load while indexing large filesystems.

```
# duc --version
duc version: 1.4.4
options: cairo x11 ui tokyocabinet
```
